kaiax/builder/impl/tx_pool_wrapping.go

```diff
@@ -81,37 +82,84 @@ func (b *BuilderWrappingModule) IsReady(txs map[uint64]*types.Transaction, next
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	tx, isReady := txs[next]
-	if b.txPoolModule != nil {
-		isReady = b.txPoolModule.IsReady(txs, next, ready)
+	// false if tx is not in txs
+	tx, ok := txs[next]
+	if !ok {
+		return false
+	}
+
+	// false if tx is not ready in child module
+	if b.txPoolModule != nil && !b.txPoolModule.IsReady(txs, next, ready) {
+		return false
 	}
-	if isReady && b.txBundlingModule.IsBundleTx(tx) {
-		if _, ok := b.knownTxs[tx.Hash()]; !ok {
-			b.knownTxs[tx.Hash()] = txAndTime{
-				tx:   tx,
-				time: time.Now(),
+
+	// add tx to knownTxs if it is a bundle tx and not in knownTxs
+	if b.txBundlingModule.IsBundleTx(tx) && !b.knownTxs.has(tx.Hash()) {
+		isNotBundleTxTail := len(ready) == 0 || !b.txBundlingModule.IsBundleTx(ready[len(ready)-1])
```

Comment:
Could you give an example where `isNotBundleTxTail` is false so the max check is skipped?
For example, given `txs: (0: A1, 1: B1, 2: T1, 3: A2, 4: B2)` and max=3, `IsReady(txs, 3, ready=[A1, B1, T1])` is false, so `IsReady(txs, 4, ready)` will not even be called. I'm not sure in which case `isNotBundleTxTail` check is effective.

Reply:
Given `txs: (0: A1, 1: B1, 2: C1, 3: T1, 4: T2)` and max=3, `IsReady(txs, 1, ready=[A1])` and `IsReady(txs, 2, ready=[A1, B1])` can return true without the max check, where (A1, B1, C1) is pair. This is just optimization.

Reply:
No C1 is not a pair. A1B1 and A2B2 are bundles.

Reply:
I meant we assume (A1, B1, C1) is bundle pair. But in your situation, only `IsReady(txs, 1, ready=[A1])` can return true without the max check.

Reply:
Oh I see. But then in this case `txs: (0: A1, 1: B1, 2: C1, 3: A2, 4: B2, 5: C2)`, everything will be added.
I think in this case we shouldn't add anything because we can't distinguish that [A1, B1, C1] and [A2, B2, C2] are different bundles. What do you think?

Reply:
if `txs: (0: A1, 1: B1, 2: C1, 3: A2, 4: B2, 5: C2)`, anything can't be added because `IsReady(txs, 0, ready=[])` is false and subsequent functions won't be called.

Reply:
Oh right, thanks. Can we make it clear that this flag is for early-return optimization?

```suggestion
		// If prev tx is bundle tx, there's no need to check the knownTxs limit because it has been checked in the previous `IsReady()` execution.
		isPrevTxBundleTx := len(ready) != 0 && b.txBundlingModule.IsBundleTx(ready[len(ready)-1])
		if  isPrevTxBundleTx {
				b.knownTxs.add(tx)
				return true
		}
```

Reply:
https://github.com/kaiachain/kaia/pull/328/commits/d91f6b5ab2006304a5fc1d4fd1902ca91f1e5dc7